### PR TITLE
feat: Support for Vitest

### LINKED ISSUE #11: Support for vitest
https://vitest.dev/ is a test framework which we need support for in the Node agent.

### PR DESCRIPTION
**Pull request by zermelo-wisen**
Fixes #11

### CODE REVIEW

**Review by dividedmind — CHANGES_REQUESTED** [review]
There are a couple of issues that I would like to be fixed.

Also, please put ESM support in a separate commit (and add tests for it, both for plain js and for typescript).

File: src/hooks/vitest.ts [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+import { args as args_, assignment, call_, identifier, literal, member, ret } from "../generate";
+import { info, warn } from "../message";
+import { recording, start } from "../recorder";
+import Recording from "../Recording";
+import genericTransform from "../transform";
+
+function createInitChannel() {
+  return new worker.BroadcastChannel("appmap-node/vitest/initialized");
+}
```
Comment (dividedmind) [inline src/hooks/vitest.ts:16]: Why this function? Seems like you should be able to just do `const initChannel = new worker.BroadcastChannel(...)` on the toplevel. Or will it lead to hanging (which is kinda weird, but if so, ok)?

File: src/hooks/vitest.ts [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    worker.isMainThread && process.argv.length > 1 && process.argv[1].endsWith("child.js");
+  const [major, ,] = process.versions.node.split(".").map(Number);
+  // There will be 2 threads (#0, #1) in [--no-threads, node 20], so we don't want
+  // to listen in thread #1 for other modes in this case. On the other hand,
+  // we don't have thread #2 in [--single-thread node 18/16], but we can listen
+  // in thread #1 because we don't have #1 in [--no-threads node 18/16].
+  // Thus we guarantee that a single thread listens in all node versions and thread modes combinations.
+  const threadIdToListenInOtherModes = major >= 20 ? 2 : 1;
+  return isNoThreadsModeMainThread || worker.threadId === threadIdToListenInOtherModes;
+}
```
Comment (dividedmind) [inline src/hooks/vitest.ts:38]: Ouch! I wish this hack wasn't necessary. Great job figuring it out and thank you for detailed comments! Fingers crossed it doesn't break anytime soon :)

File: src/register.ts [hunk 3]
```diff
@@ -23,3 +24,9 @@ Module.prototype._compile = function _compile(code: string, fileName: string): s
 Module.prototype.require = new Proxy(Module.prototype.require, {
   apply: requireHook,
 });
+
+declare global {
+  // eslint-disable-next-line no-var
+  var AppMapRecordHook: typeof recordHook;
+}
+global.AppMapRecordHook = recordHook;
```
Comment (dividedmind) [inline src/register.ts:32]: If the whole machinery in `hooks/index.ts` is now unused you can remove it.

File: src/transform.ts [hunk 4]
```diff
@@ -47,3 +52,21 @@ function dumpTree(xformed: ESTree.Program, url: URL) {
   writeFileSync(path, JSON.stringify(xformed, null, 2));
   treeDebug("wrote transformed tree to %s", path);
 }
+
+// HACK: In node 18/16, when using --loader node-ts/esm
+// sourcemap gets inserted twice to the typescript file. We remove the second one
```
Comment (dividedmind) [inline src/transform.ts:57]: This is weird. Probably there's a better way to handle this (and somehow avoid double sourcemap, which suggests the source gets transpiled twice), but I suppose it's okay for now.

File: src/bin.ts [hunk 5]
```diff
@@ -49,6 +68,10 @@ function isScript(arg: string) {
   }
 }
 
+function isESM(arg: string) {
+  return arg.match(/.*\.m[tj]s/);
```
Comment (dividedmind) [inline src/bin.ts:72]: This should also check `package.json` for `type: module`.

**Review by dividedmind — CHANGES_REQUESTED** [review]
Almost there! Great work on all the corner cases, thanks :)

File: src/register.ts (hunk 3 above)
Comment (zermelo-wisen) [reply src/register.ts:32]: Only jest hook (`hooks/jest.ts`) is using it for now. I can change it as well, if it's ok for you. Then I can remove the old machinery.

File: src/hooks/vitest.ts (hunk 1 above)
Comment (zermelo-wisen) [reply src/hooks/vitest.ts:16]: Yes, exactly, it should be called only from the right threads.

File: src/register.ts (hunk 3 above)
Comment (dividedmind) [reply src/register.ts:32]: Ah, ok. Let's leave it for now in that case.

**Review by dividedmind — APPROVED** [review]

**Pull request merged**